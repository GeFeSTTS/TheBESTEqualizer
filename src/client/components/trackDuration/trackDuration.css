.DurationContainer {
    background: var(--headerBackground-dark-super);
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    width: 100%;
    max-width: 1024px;
    margin: 0 auto;
    margin: auto;
    padding: 0 25px;
    color: var(--authFont-white);
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-box-align: center;
    -ms-flex-align: center;
}

.DurationContainerFlexChild {
    display: flex;
    align-items: center;
}

.Timer {
    width: 80px;
}


.DurationContainer--slider {
    width: 550px;
    height: 4px;
    margin: 20px;
}

.DurationContainer--slider .rangeslider__fill {
    background-color: var(--rangeSlider-ocean);
}

.DurationContainer .DurationContainerFlexChild .rangeslider__handle {
    background-color: var(--mainText-accent-blue);
    width: 20px;
    height: 20px;
    border-radius: 50%;
    border: none;
    outline: none;
}

.DurationContainer--slider .rangeslider__handle .rangeslider__handle-tooltip {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    -webkit-touch-callout: none;
}

.DurationContainer--slider .rangeslider__handle:after {
    display: none;
}

@media only screen and (max-width: 920px) {
    .DurationContainer--slider {
        width: 350px;
    }
}

@media only screen and (max-width: 720px) {
    .DurationContainer {
        flex-direction: column;
    }
}



