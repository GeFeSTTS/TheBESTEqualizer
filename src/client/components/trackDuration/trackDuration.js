import React, { Component } from 'react';
import 'react-rangeslider/lib/index.css';
import './trackDuration.css';
import PropTypes from 'prop-types';
import Slider from 'react-rangeslider';
import { connect } from 'react-redux';
import VolumeComponent from '../volumeComponent/volumeComponent';
import SwitcherSound from '../switcherSound/switcherSound';
import { playPauseSoundFromFile } from '../../store/actions/audioActions';

export class TrackDuration extends Component {
    state = {
      duration: null,
      currentTime: null,
      playing: false,
      startPlayTime: null,
      onToggle: true,
    }

    static getDerivedStateFromProps(props, state) {
      const { audioData: { loading, sound, trackName } } = props;
      if (sound && sound.sourceNode) {
        if (!state.startPlayTime && sound.playing) {
          return {
            startPlayTime: new Date(),
            duration: Math.round(sound.sourceNode.buffer.duration),
            playing: sound.playing,
            trackName,
          };
        }
        if (loading) {
          return {
            playing: false,
          };
        }
        return {
          duration: Math.round(sound.sourceNode.buffer.duration),
          playing: sound.playing,
          trackName,
        };
      }
      return null;
    }

    componentDidUpdate(prevProps, prevState) {
      if (!prevState.playing && this.state.playing) {
        this.setState({ startPlayTime: new Date(new Date() - prevState.currentTime * 1000) });
      }
      if (this.state.startPlayTime) {
        this.calculateCurrentTime();
      }
      if (prevState.trackName !== this.state.trackName) {
        this.setState({ startPlayTime: new Date() });
      }
    }

    formatingSongTime = (numberOfSecs) => {
      const seconds = numberOfSecs % 3600;
      let min = (Math.floor(seconds / 60)).toFixed();
      let sec = (seconds % 60).toFixed();

      min = String(min).padStart(2, '0');
      sec = String(sec).padStart(2, '0');
      return `${min}:${sec}`;
    }

    calculateCurrentTime = () => {
      const currentTime = new Date();
      const currentDifference = Math.round(
        (currentTime.getTime() - this.state.startPlayTime.getTime()) / 1000,
      );
      const { audioData: { sound }, playPauseSoundFromFileAsProp } = this.props;
      if (currentDifference !== this.state.currentTime && this.state.playing) {
        this.setState({ currentTime: currentDifference });
      } else if (this.state.currentTime >= this.state.duration) {
        sound.stop();
        playPauseSoundFromFileAsProp();
        this.setState({
          currentTime: 0,
          playing: false,
        });
      } else {
        setTimeout(this.calculateCurrentTime, 1000);
      }
    }

    setPauseOnSound = () => {
      const { audioData: { sound } } = this.props;
      if (!sound.paused) {
        sound.pause();
        this.setState({
          playing: false,
          onToggle: false,
        });
      }
    }

    setPlayOnSound = () => {
      const { audioData: { sound } } = this.props;
      if (!this.state.onToggle) {
        sound.play();
        this.setState({
          playing: true,
          onToggle: true,
        });
      }
    }

    changeCurrentTimeOfSong = (time) => {
      const { audioData: { sound } } = this.props;
      if (sound) {
        sound.offsetTime = time;
      }
    }

    setCurrentTime = (currentTime) => {
      this.setState({
        currentTime,
      }, () => this.changeCurrentTimeOfSong(currentTime));
    }

    render() {
      const { currentTime, duration } = this.state;
      const { audioData: { sound, voice } } = this.props;
      const minSliderVolume = 0;
      const stepSliderVolume = 0.001;
      let formatedDurationTime = '00:00';
      let formatedCurrentTime = '00:00';
      formatedDurationTime = this.formatingSongTime(duration) || formatedDurationTime;
      formatedCurrentTime = this.formatingSongTime(currentTime) || formatedCurrentTime;
      return (
        <div className="DurationContainer">
          <div className="DurationContainerFlexChild">
            <div className="Timer">
              {formatedCurrentTime}
              {' '}
/
              {formatedDurationTime}
            </div>
            <Slider
              className="DurationContainer--slider"
              value={currentTime}
              tooltip={false}
              min={minSliderVolume}
              max={duration}
              step={stepSliderVolume}
              onChangeStart={sound && this.setPauseOnSound}
              onChange={sound && this.setCurrentTime}
              onChangeComplete={sound && this.setPlayOnSound}
            />
          </div>
          {sound && (voice.playing && sound.playing) ? (
            <SwitcherSound />) : (<VolumeComponent />)}
        </div>
      );
    }
}

TrackDuration.propTypes = {
  audioData: PropTypes.instanceOf(Object).isRequired,
  playPauseSoundFromFileAsProp: PropTypes.instanceOf(Object).isRequired,
};

const mapStateToProps = state => ({
  audioData: state.audioData,
});

export default connect(mapStateToProps, {
  playPauseSoundFromFileAsProp: playPauseSoundFromFile,
})(TrackDuration);
