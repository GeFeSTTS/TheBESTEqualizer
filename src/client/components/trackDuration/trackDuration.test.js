import React from 'react';
import { shallow } from 'enzyme';
import { TrackDuration } from './trackDuration';

describe('TEST DURATION SLIDER', () => {
  let TrackDurationWrapper;
  const props = {
    tooltip: false,
    min: 0,
    max: 1,
    step: 0.001,
    audioData: {
      sound: {
        pause: () => (
          'fakePauseMethod'
        ),
        play: () => (
          'fakePlayMethod'
        ),
        stop: () => 'fakePauseMethod',
      },
      voice: {
        pause: () => (
          'fakePauseMethod'
        ),
        play: () => (
          'fakePlayMethod'
        ),
        stop: () => 'fakePauseMethod',
      },
    },
    onChangeStart: () => {},
    onChange: () => {},
    onChangeComplete: () => {},
    value: 0,
  };

  beforeEach(() => {
    TrackDurationWrapper = shallow(<TrackDuration {...props} />);
  });

  it('is render component', () => {
    expect(TrackDurationWrapper.find('.DurationContainer')).toHaveLength(1);
  });

  it('render component correctly', () => {
    expect(TrackDurationWrapper).toMatchSnapshot();
  });

  it('should call function after change', () => {
    const instance = TrackDurationWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'setCurrentTime');

    TrackDurationWrapper.setState({ currentTime: 200 });
    TrackDurationWrapper.find('Slider').simulate('change', 300);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should song play again if onToggle will false', () => {
    const instance = TrackDurationWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'setPlayOnSound');

    TrackDurationWrapper.setState({ onToggle: false, currentTime: 200, duration: 300 });
    TrackDurationWrapper.find('Slider').simulate('changeComplete', 300);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should song pause if onToggle will true', () => {
    const instance = TrackDurationWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'setPauseOnSound');

    TrackDurationWrapper.setState({ onToggle: true, currentTime: 200, duration: 300 });
    TrackDurationWrapper.find('Slider').simulate('changeStart', 250);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should song stop if current time will more then duration', () => {
    TrackDurationWrapper.setProps({ currentDifference: 100, currentTime: 200 });
    const instance = TrackDurationWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'calculateCurrentTime');

    TrackDurationWrapper.setState({ duration: 100, playing: true });
    TrackDurationWrapper.find('Slider').simulate('change', 1);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should duration from audio Data will assign in state', () => {
    TrackDurationWrapper.setProps({
      audioData: {
        sound: {
          sourceNode: {
            buffer: {
              duration: 200,
            },
          },
        },
        voice: {},
      },
    });
    TrackDurationWrapper.setState({ startPlayTime: new Date(), loading: false });

    expect(TrackDurationWrapper.state().duration).toEqual(200);
  });

  it('should time start from zero if we load another song', () => {
    const instance = TrackDurationWrapper.instance();
    const spySwitching = jest.spyOn(instance, 'calculateCurrentTime');

    TrackDurationWrapper.setState({ currentTime: 200, duration: 100, playing: true });
    TrackDurationWrapper.setProps({ currentDifference: 100 });
    TrackDurationWrapper.find('Slider').simulate('change', 1);

    expect(spySwitching).toHaveBeenCalled();
  });

  it('should startPlayTime setting again if song will changed', () => {
    TrackDurationWrapper.setState({ trackName: 'Song1', currentTime: 0, duration: 200 });
    TrackDurationWrapper.setState({
      trackName: 'Song2', currentTime: 0, duration: 200, timeForTest: new Date(),
    });

    expect(TrackDurationWrapper.state().startPlayTime).toBeTruthy();
  });
});
