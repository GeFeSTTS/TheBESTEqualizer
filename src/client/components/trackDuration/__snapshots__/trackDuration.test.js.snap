// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST DURATION SLIDER render component correctly 1`] = `ShallowWrapper {}`;
